src/Utilities/FlowAnalysis/Extensions/ListExtensions.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Analyzer.Utilities.Extensions
+{
+    internal static class ListExtensions
+    {
+        /// <summary>
+        /// Extract and remove all elements from <paramref name="list"/> which are matched by
+        /// <paramref name="predicate"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of element in the list.</typeparam>
+        /// <typeparam name="TArg">A state argument to pass to <paramref name="predicate"/>.</typeparam>
+        /// <param name="list">The list.</param>
+        /// <param name="predicate">A predicate matching elements to remove from <paramref name="list"/>.</param>
+        /// <param name="argument">An additional state argument to pass to <paramref name="predicate"/>.</param>
+        /// <returns>A collection of elements removed from <paramref name="list"/>, in the order they were removed. If
+        /// no elements were removed, this method returns <see cref="ImmutableArray{T}.Empty"/>.</returns>
+        public static ImmutableArray<T> Extract<T, TArg>(List<T> list, Func<T, TArg, bool> predicate, TArg argument)
```

Comment:
nit: can we make this an extension method?

Reply:
Done